# ተግባራት (ፈንክሽንስ)

ተግባር ወይም ፈንክሽን ተደጋጋሚ የሆኑ ስራዎችን ለማቀላጠፍ ጥቅም ላይ የሚውሉ ፕሮግራሞች ናቸው። የፓይተን ተግባራት በሦስት ዋና ዋና ክፍሎች ይከፈላሉ። እነዚህም፡ 

* የውስጥ ስሬት የሆኑ (ቢዩልት ኢን) የምንላቸው። እነዚህም እስካሁን በጻፍናቸው ፕሮግራሞች ውስጥ ስንጠቀማቸው የነበሩ ናቸው። **print()** እና **type()** እንደምሳሌ መጥቀስ ይቻላል። ተጨማሪ የውስጥ ስሬት የሆኑ ተግባራትን [እዚህ](https://docs.python.org/3/library/functions.html) መመልከት ይቻላል።
* ተጠቃሚው የሚጽፋቸው (ዩዘር ዲፋይንድ) የምንላቸው። የፓይተን ተጠቃሚዎች (ፕሮግራመሮች) የሚጽፏቸው ናቸው። እነዚህን በዚህ ክፍል እንመለከታለን። 
* ስም አልባ (አኖኒመስ) የምንላቸው።  **lambda**ን እንደምሳሌ መጥቀስ ይቻላል። እንዲህ አይኖቶች ተግባራትን ለማወጅ የምንጠቀመውን ቁልፍ ቃል **def** የማይጠቀሙ፤ ነገር ግን እንደተግባር ተደጋጋሚ ስራዎችን የሚያቀላጥፉ ናቸው። 


## ተግባርን ማወጅ

ተግባርን ለማወጅ የሚከተለውን መንገድ መጠቀም እንችላለን። 
* **def** የሚለውን ቁልፍ ቃል በመጠቀም መጀመር እና የተግባሩን ስም ማስከተል ነው። 
* ልኬቶችን (ፓራሜትሮችን) ካሉን ከተግባሩ ስም በመቀጠል በቅንፍ ውስጥ ማስቀመጥ። ልኬቶች ከሌሉ ባዶ ቅንፍን አስቀምጦ ሁለት ነጥቦችን (ኮለን) ማድረግ። 
* ተግባሩ ሊከውናቸው የሚገቡ መግለጫዎችን (ስቴትመንቶችን) ማስከተል። 
* በመጨራሻም ተግባሩ ስራውን ሰርቶ የሚመልሰውን ወይም በሚያትመው ነገር ማብቃት። 

በምሳሌ እንመልከት
***
``` 
def ሰላምታ ():
    print("ሰላም ለሁላችን።")
```
***

ከላይ የተጻፈው የተግባርን ብያኔ ያሟላ ነው። ይህም `def` የሚል የተግባር ማወጃ ቁልፍ ቃል፣ `ሰላምታ` የሚል የተግባር ስም፣ ነጋሪ እሴቶችን (አርጉመንት) ስለማይወስድ ልኬቱ ባዶ ቅንፍ `()` እና  ስለ`:` በማስከተል ታውጇል። ተግባሩ የሚያከናውነው ስራ *ሰላም ለሁላችን* የሚል ቃል ማተም ስለሆነ አንድ መግለጫ ወይም ስቴትመንት አለው።  

In [4]:
# ተግባር - ምሳሌ አንድ 
# የተግባሩ ስም ሰላምታ 
# ይተግባሩ አላም ሰላም ለሁላችን የሚል መልክት ማሳየት 
# ተግባሩን ካሮጥን ቦኋላ ውጤቱን ለማየት ሰላምታ() በማለት ተግባሩን መጥራት 
def ሰላምታ ():
    print("ሰላም ለሁላችን።")
    
# ተግባሩን መጥራት 
ሰላምታ()

ሰላም ለሁላችን።


## ተግባርን መጥራት

ተግባርን መጥራት ማለት የሚፈለገውን ስራ ለማከናወን የሚረዳውን ተግባር በተልይ መጥራት ማለት ነው። ከላይ የተመለከትነው  `ሰላምታ` ተግባር *ሰላም ለሁላችን* የሚል መልእክት ማሳየት ነው። ይህንን መልእክት ማሳየት በፈለግን ጊዜ መልእክቱን ከመጻፍ ተግባሩን በስሙ በፕሮግራማችን መሀል መጥራት እንችላለን ማለት ነው። በምሳሌ እንመልከት፡ 

In [7]:
def ሰላምታ ():
    print("ሰላም ለሁላችን።")
    
# ሰላምታን መጥራት 
ሰላምታ()
# ቁጥርን በራሱ የሚያባዛ የብዜት ሰንጠረዥ ተደጋጋሚ 
for i in range(5):
    print(i, "\t", i*i)
# ሰላምታን ድጋሚ መጥራት 
ሰላምታ()
# ቁጥርን በራሱ የሚደምር
for i in range(5):
    print(i, "\t", i+i)
# ሰላምታን ለሦስተኛ ግዜ መጥራት 
ሰላምታ()

ሰላም ለሁላችን።
0 	 0
1 	 1
2 	 4
3 	 9
4 	 16
ሰላም ለሁላችን።
0 	 0
1 	 2
2 	 4
3 	 6
4 	 8
ሰላም ለሁላችን።


## መላሽ ተግባራት

መላሽ ተግባራት ተግባራቸውን አከናውነው ሲጨሩሱ ውጤታቸውን ለጠሪያቸው የሚመልሱ ማለት ነው። እነዚህም `return` የሚለውን ቁልፍ ቃል በመጨረሻ በመጠቀማቸው ይለያሉ።   

In [9]:
# መላሽ ተግባር ምሳሌ 
# ሁለት ቁጥሮችን የሚደምር ተግባር
# ልኬት (ፓራሜትር) ሀ እና ለ 
# መላሽ ተግባር መሆኑን return የሚለውን ቁልፍ ቃል ተመልከቱ 
def ደምር (ሀ, ለ):
    """ ሁለት ቁጥሮችን የሚደምር ተግባር። 
    ለመጠቀም ደምር(5, 10) ብለን ብንጠራው ውጤቱ 15 ይሆናል ማለት ነው።"""
    return ሀ+ለ 

# ተግባሩን መጥራት 
ደምር(10.50, 12.75)

23.25

In [27]:
# ተጨማሪ የመደመር ምሳሌ
# በተመሳሳይ ስፍራ ያሉ ቁጥሮችን መደመር 
# ሀ= (1,2,3,4) እና ለ= (3, 9, 6, 7) ቢሰጡን 
# (1+3, 2+9, 3+6, 4+7) መደመር 
# ይሄን ለማከናወን ፓይተን zip() የሚባል የውስጥ ስሬት ተግባር አለው። 
# ዚፕ ሀ እና ለ ን ወስዶ (1 ,3) (2 ,9) (3 ,6) (4 , 7) አድርጎ የመልሳቸዋል። 
# ከ ዚፕ የተመለሰውን የሚፈለገውን ቀመር ማካሄድ እንችላለን 
def ስፍራን_ደምር(ሀ, ለ):
    """ተመሳሳይ ስፍራ ላይ ያሉ ቁጥሮችን መደመር"""
    return [ሀ + ለ for ሀ, ለ in zip(ሀ, ለ)]

In [28]:
ሀ = (1,2,3,5)
ለ = (3,4,5,6)
ስፍራን_ደምር(ሀ, ለ)

[4, 6, 8, 11]

መቀነስንም ሆነ ማባዛትን በተመሳሳይ መልኩ መስራት እንችላለን። 

## ተጨማሪ የተግባር ምሳሌዎች

ከላይ የተመለከትነው `ስፍራን_ደምር` የሚለው ተግባር አንድ ሳንካ አለበት እሱም ልኬቱ (ፓራሜትሩ) ነው። ከሁለት በላይ የእሴት ነጋሪዎችን መቀበል አይችልም። 

*** 
ስፍራን_ደምር የሚለውን ተግባር ሶስትና ከዛ በላይ የሆኑ እሴት ነጋሪዎችን በመጠቀም ስፍራዎችን እንዲደምር አድርጉት። ማለትም፡ 
```
ሀ = (1,2,3,5)
ለ = (3,4,5,6)
መ = (3,4,5,6)
ሰ = (3,4,5,6)
ረ = (3,4,5,6)
ወስዶ የሚመልሰው ውጤት [13, 18, 23, 29] እንዲሆን ማለት ነው። 
```
በራሳችሁ ሞክሩትና አማራጭ መልሱን ከታች ተመልከቱ። 
***

In [30]:
#ሁለት እና ከዛ በላይ የሆኑ ነጋሪ እሴቶችን መቀበል የሚችል ስፍራን_ደምር ተግባር 
# ግባት፡ 
#ሀ = (1,2,3,5)
#ለ = (3,4,5,6)
#መ = (3,4,5,6)
#ሰ = (3,4,5,6)
#ረ = (3,4,5,6)
# ውጤት፡ [13, 18, 23, 29]
# መፍትሄ 
# መጀመሪያ ሀ ን ለብቻው በተለዋዋጭ ላይ ማስቀመጥ
# በመቀጠልም ከ ለ እስከ መጨረሻው ያሉትን አንድ በአንድ ተደጋጋሚ በመጻፍ እየወሰዱ ለ ስፍራን_ደምር ማስተላለፍ

def ስፍራን_ደምር(ሀ, ለ):
    """ተመሳሳይ ስፍራ ላይ ያሉ ቁጥሮችን መደመር"""
    return [ሀ + ለ for ሀ, ለ in zip(ሀ, ለ)]


def ስፍራን_ሁሉም(ሁሉም):
    """ሁሉንም ተመሳሳይ ስፍራ ያሉ ቁጥሮችን መደመር"""
    ውጤት = ሁሉም[0] # ይሄ ሀን ነጥሎ ይወስዳል
    for ሁሉ in ሁሉም[1:]: # ሁሉንም አንድ በአንድ እየወሰድ ሁሉ ለሚለው ተለዋዋጭ ይሰጠዋል
        ውጤት = ስፍራን_ደምር(ውጤት, ሁሉ) # ስፍራ_ደምር ተግባርን በመጠቀም ውጤት ላይ መደመር። 
    return ውጤት

ሀ = (1,2,3,5)
ለ = (3,4,5,6)
መ = (3,4,5,6)
ሰ = (3,4,5,6)
ረ = (3,4,5,6)
አ=(ሀ,ለ,መ,ሰ,ረ)
ስፍራን_ሁሉም(አ)

[13, 18, 23, 29]

## መሰረታዊ ስታትስቲክስ

አጠቃላይ የፓይተንን የተለያዩ ጥቅሞች ለማሰየት እና በዛውም ፕሮግራምን በተግባር እየከፋፈሉ መጻፍን የበለጠ ለመረዳት ይረዳ ዘንድ ያደረኩት ዳሰስ እንጂ ፓይተን እጅግ የላቁ ብዙ ዘመናዊ ጥቅሞች ያሉት የኮምፒውተር ቋንቋ ነው። በዚህ ክፍል መሰረታዊ ስታትስቲክስን እንዴት ፓይተንን በመጠቀም እንደምንጽፍ እንመለከታለን። በዋናነት የምንመለከተው የአንድን ዳታ አማካኝ እንዴት ማግኘት እንደምንችል ነው። ከላይ ለመጥቀስ እንደሞከርኩት አላማው ተግባርን ይበልጥ ለመረዳት እንጂ ስታትን ማሳወቅ አይደለም። ስታትስቲክስ በፓይተን እራሱን የቻለ ሌላ ትልቅ መጽሀፍ ሊወጣው የሚችል አርእስት ነውና።  

### አማካኝ (ሚን) 

አማካኝ ወይም ሚን የውሂባችንን አማካኝ እሴት ለመገመት ይረዳል። ለምሳሌ አንድ በአዲስ አበባ የሚኖር የሚኒባስ ሾፌር ቅዳሜ ቀን በአማክኝ 438.92 ብር ያስገባል ብንል፤ ከተለያዩ ሾፌሮች የተለያዩ ቦታዎች ላይ በመገኘት አራት ቅዳሜዎችን ለዳሰሳ ጥናት በመሰብሰብ አማካኝ ገቢውን በቁጥር ማስቀመጥ ይቻላል። 

ከላይ የጻፍነውንም ፕሮግራም ጥቅም ላይ የሚያውል እንዲሆን የሚከተለውን ውሂብ እንመልከት። 2 ፈቃደኛ የሆኑ የሚኒባስ ሾፌሮችን አራት የተለያዩ ቦታዎች (አራት ኪሎ፣ ሜክሲኮ፣ መርካቶ እና ፒያሳ) በዚሁ ቅደም ተከተል መነሻቸውን አድርገው በሚሰሩበት ቀን ለሁለት ቅዳሜዎች ወደቤታቸው በሚገቡበት ሰዓት የሚከተለውን ውሂብ ሰበሰብን እንበል።

*** 
ሾፌር አንድ፡- 
```
የመጀመሪያ ቅዳሜ [659.20, 334.75, 522.50, 327.50] 
ሁለተኛ ቅዳሜ [569.35, 522.75, 318.80, 439.90]
```
ሾፌር ሁለት፡-
```
የመጀመሪያ ቅዳሜ [418.30, 294.15, 691.85, 217.65] 
ሁለተኛ ቅዳሜ [299.20, 522.15, 381.55, 225.45] 
```
*** 
በስብር ቅንፍ ውስጥ ያለው ውሂብ ቅደም ተከተሉ \[አራት ኪሎ, ሜክሲኮ, መርካቶ, ፒያሳ\] ነው። 

In [54]:
# ውሂቡ በቂ ባይሆንም ለምሳሌ ያህል ነው። 
#አንድ ሾፌር አራት ኪሎ ላይ በአማካኝ ስንት ብር ያገኛል ብንል። 
# መጀመሪያ አንድ ሾፌር አራት ኪሎ, ሜክሲኮ, መርካቶ, ፒያሳ በተናጠል ስንት እንደሚገኝ መደመር አለብን 
# ይህን ለማድረግ ስፍራን_ደምር በማሻሻል መጠቀም እንችላለን። 
def የስፍራ_አማካኝ(ሀ, ለ):
    """ተመሳሳይ ስፍራ ላይ ያሉ ቁጥሮችን መደመር"""
    return [ሀ + ለ for ሀ, ለ in zip(ሀ, ለ)]

def አማካኝ_የሁሉም(ሁሉም):
    """ሁሉንም ተመሳሳይ ስፍራ ያሉ ቁጥሮችን መደመር"""
    ውጤት = ሁሉም[0] 
    ቆጣሪ=1 # የስፍራ_አማካኝ በተጠራ ቁጥር እየጨመረ ይሄዳል። አማካኝ ግባቶች በርዝማኔ መካፈል ስላለባቸው። 
    for ሁሉ in ሁሉም[1:]: 
        ውጤት = የስፍራ_አማካኝ(ውጤት, ሁሉ) 
        ቆጣሪ+=1
    return [አጠቃላይ_ድምር/ቆጣሪ for አጠቃላይ_ድምር in ውጤት]

# ሾፌር አንድ 
ቅዳሜ፩_ሾፌር_አንድ= (659.20, 334.75, 522.50, 327.50) 
ቅዳሜ፪_ሾፌር_አንድ= (569.35, 522.75, 318.80, 439.90)

# ሾፌር ሁለት 
ቅዳሜ፩_ሾፌር_ሁለት = (418.30, 294.15, 691.85, 217.65) 
ቅዳሜ፪_ሾፌር_ሁለት = (299.20, 522.15, 381.55, 225.45)

# አጠቃላዩን ሰብስቦ ለ አማካኝ_የሁሉም ለማሳለፍ 
አማካኝ = (ቅዳሜ፩_ሾፌር_አንድ, ቅዳሜ፪_ሾፌር_አንድ, ቅዳሜ፩_ሾፌር_ሁለት, ቅዳሜ፪_ሾፌር_ሁለት)
# አንድ ሾፌር አራት ኪሎ በአማካኝ የሚያገኘው የቀን ገቢው ብንደምረው 
#(659.20 + 569.35 + 418.30+299.39)/4 = 486.5375
# ፕሮግራማችንን ተጠቅመን ውጤቱን እናረጋግጥ 
አማካኝ_በሰፈር = አማካኝ_የሁሉም(አማካኝ)
# ውጤቱን መልክ ለማስያዝ ማለትም ከነጥብ ቦኋላ 2 ዲጂት እንዲያትም ለማድረግ ነው
[ '%.2f' % እሴት for እሴት in አማካኝ_በሰፈር ]

['486.51', '418.45', '478.68', '302.62']

አራት ኪሎ አካባቢ አንድ የሚኒባስ ሹፌር የቅዳሜ ገቢው በአማካኝ 486.51 ሳንቲም ይሆናል ማለት ነው። በሚቀጥለው ክፍል የአማርኛ ጽሁፎችን እንዴት እንደምንተነትን እንመለከታለን። 